File: 백준/Silver/14501. 퇴사/퇴사.py
import sys
input = sys.stdin.readline
N = int(input())
arr = [list(map(int,input().split())) for _ in range(N)]
ans = 0
def dfs(n,sm):
    global ans
    if n == N:
        if ans < sm:
            ans = sm
        return

    if n+arr[n][0] <= N:
        dfs(n+arr[n][0],sm+arr[n][1])
    dfs(n+1,sm)
dfs(0,0)
print(ans)

File: 백준/Silver/14501. 퇴사/퇴사.js
let input = require("fs").readFileSync("dev/stdin").toString().trim().split("\n");
const N = Number(input[0]);
const lst = input.slice(1, N + 1).map((el) => el.split(" ").map(Number));
let answer = 0;

function dfs(n) {
  if (n > N) {
    return -123512351233624;
  }

  if (n === N) {
    return 0;
  }

  return Math.max(dfs(n + lst[n][0]) + lst[n][1], dfs(n + 1));
}

answer = dfs(0);
console.log(answer);
